src/Compilers/CSharp/Test/Symbol/Symbols/PartialPropertiesTests.cs

```diff
@@ -2213,18 +2213,613 @@ class C
+                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(10, 28),
+                // (11,23): error CS8988: The 'scoped' modifier of parameter 'i' doesn't match partial definition.
+                //     public partial RS this[scoped ref int i, int _] { get => default; set { } }
+                Diagnostic(ErrorCode.ERR_ScopedMismatchInParameterOfPartial, "this").WithArguments("i").WithLocation(11, 23),
+                // (11,35): error CS0631: ref and out are not valid in this context
+                //     public partial RS this[scoped ref int i, int _] { get => default; set { } }
+                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(11, 35));
+        }
+
+        [Fact]
+        public void Semantics_OptionalParameters()
+        {
+            var source = """
+                using System;
+
+                var c = new C();
+                Console.Write(c[1]);
+                Console.Write(c[1, 2]);
+
+                partial class C
+                {
+                    public partial int this[int x, int y = 1] { get; }
+                    public partial int this[int x, int y] { get => y; }
+                }
+                """;
+
+            var verifier = CompileAndVerify(source, expectedOutput: "12");
+            verifier.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void OptionalParameters_OnImplementationPart()
+        {
+            // A warning is reported for optional parameters on implementation part, even if it matches the definition part.
+            var source = """
+                partial class C
+                {
+                    public partial int this[int x, int y] { get; set; }
+                    public partial int this[int x, int y = 1] { get => y; set { } }
+
+                    public partial int this[int x, int y, int z = 1] { get; set; }
+                    public partial int this[int x, int y, int z = 1] { get => z; set { } }
+                }
+                """;
+
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (4,40): warning CS1066: The default value specified for parameter 'y' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
+                //     public partial int this[int x, int y = 1] { get => y; set { } }
+                Diagnostic(ErrorCode.WRN_DefaultValueForUnconsumedLocation, "y").WithArguments("y").WithLocation(4, 40),
+                // (7,47): warning CS1066: The default value specified for parameter 'z' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
+                //     public partial int this[int x, int y, int z = 1] { get => z; set { } }
+                Diagnostic(ErrorCode.WRN_DefaultValueForUnconsumedLocation, "z").WithArguments("z").WithLocation(7, 47));
```

Comment:
nit: This can be punted to a later PR, but it will be good to hunt down the other couple places that check `IsPartialImplementation()` and test them too.

Reply:
I went through this briefly and added a few prototype comments